### DISCUSSION TITLE
Creating an outline around the text created by `vips_text`

### DISCUSSION THREAD

**Discussion by kipcole9** [libvips/libvips#3108] [Help wanted]
I'm working on another challenge - to build a meme generator using libvips. 98% done and its been fun. But .... given an image created by `vips_text`, how would I go about putting an outline around the glyphs?  

The common meme generators use white text with black outlines around the glyphs. I could easily do it with SVG, but SVG word wrapping seems not to be a thing.  `vips_text` does a great job wrapping and fitting to a box - but not outlining.  I considered creating another layer of text slightly larger and composing the layers. But the different sized text would also have different glyph spacing so the effect won't align.

I looked for a way to place pixels along an edge. And then realised I have no idea how to do that. 

So my image processing education continues at Cupitt University (I hope).  Here's the example text of a typical meme as a refresher:

![image](https://user-images.githubusercontent.com/15895/197213210-650f8227-e937-4aed-b6d1-70243761059a.png)

**Comment by jcupitt** [comment]
Hi again Kip,

You can use `dilate` to expand a mask, eg.:

```python
#!/usr/bin/python3

import sys
import pyvips

text = pyvips.Image.text(sys.argv[2], dpi=600)

radius = 5

# we need to make the text image large enough to hold the extra edges
text = text.embed(radius, radius,
    text.width + 2 * radius, text.height + 2 * radius)

# dilate the text mask, then apply a slight blur to alias the
# edges
circle_mask = pyvips.Image.black(radius * 2 + 1, radius * 2 + 1) \
    .add(128) \
    .draw_circle(255, radius, radius, radius, fill=True)
wide_text = text.dilate(circle_mask).gaussblur(0.5, min_ampl=0.1)

# paint wide text blue
background_text = wide_text \
    .new_from_image([128, 128, 255]) \
    .bandjoin(wide_text) \
    .copy(interpretation="srgb")

# paint original text black
foreground_text = text \
    .new_from_image([0, 0, 0]) \
    .bandjoin(text) \
    .copy(interpretation="srgb")

# composite black over blue
image = background_text.composite(foreground_text, "over")

image.write_to_file(sys.argv[1])
```

If I run:

```
$ ./outline.py x.png "Hello, World!"
```

I get:

![x](https://user-images.githubusercontent.com/580843/197218552-109956f0-bccf-4c86-92cf-fd165ffdfe1b.png)

**Reply by jcupitt** [reply]
`dilate` is `Morph("dilate")`, in case go-vips is missing that shortcut.

**Reply by jcupitt** [reply]
... have you seen nip2? It's a libvips GUI I made a while ago:

https://github.com/libvips/nip2

You can use it to experiment with image processing pipelines before coding them up. It's a bit like a spreadsheet, so you can watch all the pixels while you adjust the parameters. Handy!

![image](https://user-images.githubusercontent.com/580843/197224790-130c090b-2154-4e8e-8677-6eec86df7a4f.png)

That workspace file is here:

www.rollthepotato.net/~john/outline.ws

**Reply by kipcole9** [reply]
I have seen NIP2, but ... I'm on a Mac and NIP2 and Mac really seem to not get on well (the downloadable binaries aren't compatible with recent releases) and compiling from source brought no joy. Perhaps I'm misusing something?

**Reply by jcupitt** [reply]
Ah perhaps I've tried to push it on you before, I don't remember. Yes, the mac build needs redoing :( I'm tinkering with a gtk4 rewrite at the moment, perhaps that'll come to macos.

**Comment by kipcole9** [comment]
Wow, fantastic, thanks very much John.  Really appreciate your engagement and support. I'm having way too much fun with `libvips` and learning a lot. Hope I can repay you and the community at some point when I get my training wheels off.

**Comment by kipcole9** [comment]
For anyone landing here, thanks to John's help meme generation made simple in the [Elixir](https://elixir-lang.org) [Image](https://github.com/kipcole9/image) library (that is based on `libvips`) using the [Image.meme/3](https://github.com/kipcole9/image/blob/main/lib/image.ex#L2025-L2083) function:
```elixir
Image.open!("./test/support/images/meme.jpg") 
|> Image.meme!("One simply cannot", text: "Enjoy image processing without libvips") 
|> Image.preview
```
<img width="279" alt="image" src="https://user-images.githubusercontent.com/15895/197318123-7a213841-06db-4d8f-9819-9efdf4bd1753.png">
